farm_ng/core/serial/proto_reader_writer.h

```diff
@@ -0,0 +1,69 @@
+// Copyright (c) farm-ng, inc.
+//
+// Use of this source code is governed by an MIT-style
+// license that can be found in the LICENSE file or at
+// https://opensource.org/licenses/MIT.
+
+#pragma once
+
+#include "farm_ng/core/logging/logger.h"
+
+#include <google/protobuf/text_format.h>
+#include <google/protobuf/util/json_util.h>
+
+#include <filesystem>
+#include <fstream>
+
+namespace farm_ng {
+
+/// TODO: Add API docs.
+
+void writeProtobufToJsonFile(
+    const std::filesystem::path& path, const google::protobuf::Message& proto);
+
+void writeProtobufToBinaryFile(
+    const std::filesystem::path& path, const google::protobuf::Message& proto);
+
+template <class ProtobufT>
+ProtobufT readProtobufFromJsonFile(const std::filesystem::path& path) {
```

Comment:
is it possible to hid the implementation of this as much as possible to avoid compiler overhead? Suggest calling a non-templated function from the templated function, that takes a ``google::protobuf::Message* out_proto`` output parameter.